Add placeholder and pattern to daterange picker widget

esoergel opened the pull request:
## Product Description
Adds a placeholder text with the expected format to the date range widget:
![image](https://user-images.githubusercontent.com/2367539/225035026-53966d8b-fbf7-4619-af0e-694da3ef4ccf.png)
It also adds loose validation so things like "Jan 1, 2022" will show up red when the user starts typing them

## Technical Summary
https://dimagi-dev.atlassian.net/browse/QA-4867

## Feature Flag
Case Search: Enable synchronous mobile searching and case claiming

## Safety Assurance

### Safety story


### Automated test coverage

n/a

### QA Plan

Grew out of QA on https://github.com/dimagi/commcare-hq/pull/32628
https://dimagi-dev.atlassian.net/browse/QA-4867

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Add expected daterange format as placeholder

esoergel committed:
Add html pattern to date range input

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -370,6 +370,9 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
                 autoUpdateInput: false,
                 "autoApply": true,
             });
+            this.ui.dateRange.attr("placeholder", dateFormat + separator + dateFormat);
```
I think we should redefine `dateFormats` to prioritize `dateFormat` in the list of allowable formats if `dateFormat` is the recommended input pattern.

This is because we parse the input of `dateRange` using  `dateFormats` as order of priority input formats. `parts = _.map(oldValue.split(separator), function (v) { return moment(v, dateFormats);`

esoergel replied:
Fair point, just updated it.  I'd also forgotten about that part that parses input formats for the date range picker.  This means we can consistently apply the input parsing to both widgets!  Working on that now

Jtang-1 replied:
That's a win for consistent date picker behavior!

I thought you were running into issues previously with using strict mode on date range picker based on this comment. But I now realize I think you were looking at something else https://github.com/dimagi/commcare-hq/pull/32628#issuecomment-1460860379

esoergel replied:
I was, because I was trying to do it in the library (as the single date picker does with `parseInputDate`).  Turns out we already interpret the dates in the date range picker in this code though, so i can just jump in there.  I did a bunch more fiddling to unify input processing.  I still need to do more testing, but let me know what you think.

esoergel committed:
Preference stated format in coercion; be strict

esoergel committed:
Extract parseInputDate for reuse and testing

esoergel committed:
Use parsed date even in special case

esoergel committed:
remove nonsensical conditional

Dates aren't moment objects, so the first part of the conditional would
be hit anyways.  Besides, I don't this happens, since
convertTwoDigitYear would fail hard on a Date input if one were passed
in

esoergel committed:
Allow formats that don't zero-pad months and days

esoergel committed:
stickler

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -380,18 +378,20 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
-                        newValue = parts[0].format(dateFormat) + separator + parts[1].format(dateFormat);
+                        [start, end] = parts;
                     }
+                    newValue = start.format(dateFormat) + separator + end.format(dateFormat);
```
nice clean up

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/utils.js:
```diff
@@ -371,13 +371,23 @@ hqDefine('cloudcare/js/utils', [
+    var dateFormats = ['MM/DD/YYYY', 'YYYY-MM-DD', 'M/D/YYYY', 'M/D/YY', 'M-D-YYYY', 'M-D-YY', moment.defaultFormat];
+
+    /** Coerce an input date string to a moment object */
+    var parseInputDate = function (dateString) {
```
I feel like something weird happens if the dateString input is `YYYY-MM-DD` since `convertTwoDigitYear ` will try to convert the DD into four digits. But the unit test seems to pass and i can't figure out why

Jtang-1 replied:
Ah, no issue here. `YYYY-MM-DD` won't be converted because `convertTwoDigitYear` checks if `dateString` has <= 6 digits before converting

esoergel replied:
Nice, yeah that makes sense - it might even be reasonable to switch to a more strict pattern like regex matching `^\d{1,2}[.-/]\d{1,2}[.-/]\d{1,2}$`.  Hell, I wonder if temporarily patching `moment.parseTwoDigitYear` might be an option if we did something like:
```js
oldParseTwoDigitYear = moment.parseTwoDigitYear;
moment.parseTwoDigitYear = function () {...};
moment.parseTwoDigitYear = oldParseTwoDigitYear
```

Anyways, I think iterating via unit tests is probably our best bet here, especially given that part of this is controlled somewhat opaquely by a library.  Once I switched to that approach, I felt a lot more confident in this code.  Honestly I should've done it much sooner, but I guess I'm just not used to that in javascript.

esoergel replied:
totally agree

Jtang-1 reviewed (commented):
defining input parsing in one spot makes a lot of sense and great change

Jtang-1 reviewed (approved):
Pulled the change in locally and tested some cases. Didn't see any issues!

orangejenny commented:
Is this going through QA?
